numpyro/hmc_util.py

```diff
@@ -781,3 +782,106 @@ def single_chain_init(key, only_params=False):
         _, potential_fn, constrain_fun = single_chain_init(rng[0])
         init_params = vmap(lambda rng: single_chain_init(rng, only_params=True))(rng)
         return init_params, potential_fn, constrain_fun
+
+
+# TODO: to be replaced by np.cov ddof=1 for the next JAX version
+def _cov(samples):
+    wc_init, wc_update, wc_final = welford_covariance(diagonal=False)
+    state = wc_init(samples.shape[1])
+    state = fori_loop(0, samples.shape[0], lambda i, state: wc_update(samples[i], state), state)
+    return wc_final(state)[0]
+
+
+def consensus(subposteriors, num_draws, diagonal=False, rng=None):
+    """
+    Merges subposteriors following consensus Monte Carlo algorithm.
+
+    **References:**
+
+    1. *Bayes and big data: The consensus Monte Carlo algorithm*,
+       Steven L. Scott, Alexander W. Blocker, Fernando V. Bonassi, Hugh A. Chipman,
+       Edward I. George, Robert E. McCulloch
+
+    :param list subposteriors: a list in which each element is a collection of samples.
+    :param int num_draws: number of draws from the merged posterior.
+    :param bool diagonal: whether to compute weights using variance or covariance, defaults to
+        `False` (using covariance).
+    :param jax.random.PRNGKey rng: source of the randomness, defaults to `jax.random.PRNGKey(0)`.
+    :return: a collection of `num_draws` samples with the same data structure as each subposterior.
+    """
+    rng = random.PRNGKey(0) if rng is None else rng
+    # stack subposteriors
+    joined_subposteriors = tree_multimap(lambda *args: np.stack(args), *subposteriors)
+    # shape of joined_subposteriors: n_subs x n_samples x sample_shape
+    joined_subposteriors = vmap(vmap(lambda sample: ravel_pytree(sample)[0]))(joined_subposteriors)
+
+    # randomly gets num_draws from subposteriors
+    n_subs = len(subposteriors)
+    n_samples = tree_flatten(subposteriors[0])[0][0].shape[0]
+    # shape of draw_idxs: n_subs x num_draws x sample_shape
+    draw_idxs = random.randint(rng, shape=(n_subs, num_draws), minval=0, maxval=n_samples)
+    joined_subposteriors = vmap(lambda x, idx: x[idx])(joined_subposteriors, draw_idxs)
```

Comment:
why should `joined_subposteriors` depend on `draw_idxs` instead of all the samples present? is this what they do in the reference?

Reply:
The paper only mentioned that we draw num_draws from each subposterior. I guess you wonder about the case these samples are already resampled, so we don’t need to do resample here? That makes sense to me. Currently, numpyro just returns all samples. I don’t need extra codes for this resampling job. So I’ll make num_draws=None by default, which only merges these subposteriors. If num_draws not None, it will do resampling job for each subposterior to get num_draws samples for my convenience.

Reply:
done!